File: scripts/backend-node/razorpay-migration/dualWriteService.js
/**
 * Dual-write canary routing - Razorpay <-> Paytm.
 *
 * Sticky hashing on customerId ensures a given customer always hits the same
 * gateway across retries within the canary window. The percentage is read
 * from env on every call so you can bump it without redeploying.
 */

import crypto from "node:crypto";
import Razorpay from "razorpay";
import PaytmChecksum from "paytmchecksum";

const RAZORPAY = new Razorpay({
  key_id: process.env.RAZORPAY_KEY_ID,
  key_secret: process.env.RAZORPAY_KEY_SECRET,
});

const PAYTM = {
  mid: process.env.PAYTM_MID,
  key: process.env.PAYTM_MERCHANT_KEY,
  websiteName: process.env.PAYTM_WEBSITE_NAME || "WEBSTAGING",
  env: process.env.PAYTM_ENVIRONMENT || "staging",
};
const PAYTM_BASE =
  PAYTM.env === "production"
    ? "https://secure.paytmpayments.com"
    : "https://securestage.paytmpayments.com";

/** Sticky 0-99 bucket per customer. */
function bucket(customerId) {
  const h = crypto.createHash("sha256").update(String(customerId)).digest();
  return h.readUInt32BE(0) % 100;
}

/** Returns "paytm" or "razorpay" for this customer. */
export function pickPsp(customerId) {
  const canary = Number(process.env.CANARY_PCT || "0");
  return bucket(customerId) < canary ? "paytm" : "razorpay";
}

/** Create order on Razorpay. Amount is INR rupees as a number. */
async function createRazorpayOrder({ orderId, amount, customerId }) {
  const res = await RAZORPAY.orders.create({
    amount: Math.round(amount * 100), // paise
    currency: "INR",
    receipt: orderId,
    notes: { customer_id: customerId },
  });
  return {
    psp: "razorpay",
    pspOrderId: res.id, // razorpay-issued
    clientPayload: { key: process.env.RAZORPAY_KEY_ID, order_id: res.id, amount: res.amount },
  };
}

/** Create order on Paytm. Amount is INR rupees as a number. */
async function createPaytmOrder({ orderId, amount, customerId }) {
  const body = {
    requestType: "Payment",
    mid: PAYTM.mid,
    websiteName: PAYTM.websiteName,
    orderId,
    callbackUrl: process.env.PAYTM_CALLBACK_URL,
    txnAmount: { value: amount.toFixed(2), currency: "INR" },
    userInfo: { custId: String(customerId).replace(/[^A-Za-z0-9_@-]/g, "_") },
  };
  const signature = await PaytmChecksum.generateSignature(JSON.stringify(body), PAYTM.key);
  const url = `${PAYTM_BASE}/theia/api/v1/initiateTransaction?mid=${PAYTM.mid}&orderId=${orderId}`;
  const r = await fetch(url, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ head: { signature }, body }),
  });
  const json = await r.json();
  if (!json?.body?.txnToken) {
    throw new Error(`Paytm initiate failed: ${json?.body?.resultInfo?.resultMsg || "no token"}`);
  }
  return {
    psp: "paytm",
    pspOrderId: orderId,
    txnToken: json.body.txnToken,
    clientPayload: {
      mid: PAYTM.mid,
      orderId,
      txnToken: json.body.txnToken,
      amount: amount.toFixed(2),
    },
  };
}

/** Top-level dispatch. Use this from your /api/checkout/start endpoint. */
export async function createOrder({ orderId, amount, customerId }) {
  const psp = pickPsp(customerId);
  // emit telemetry (metric: psp, customerId hash, amount, outcome)
  const start = Date.now();
  try {
    const result =
      psp === "paytm"
        ? await createPaytmOrder({ orderId, amount, customerId })
        : await createRazorpayOrder({ orderId, amount, customerId });
    return { ...result, latencyMs: Date.now() - start, pspRouted: psp };
  } catch (err) {
    // Optional fallback to the other PSP if one is down. Keep symmetric retry
    // logic out of this file - that's a separate concern.
    err.pspRouted = psp;
    throw err;
  }
}

File: scripts/backend-python/razorpay_migration/dual_write_service.py
"""
Dual-write canary routing - Razorpay <-> Paytm (Python / Flask).

Sticky hashing on customer_id keeps a given customer pinned to one gateway
within the canary window. CANARY_PCT is read from env on every call so you can
roll out / back without redeploying.
"""

from __future__ import annotations

import hashlib
import json
import os
from decimal import Decimal, ROUND_HALF_UP

import razorpay
import requests
from paytmchecksum import PaytmChecksum


def _money(amount) -> str:
    """Two-decimal string. Paytm rejects '1', '1.0', '1.000'."""
    q = Decimal(str(amount)).quantize(Decimal("0.01"), rounding=ROUND_HALF_UP)
    return f"{q:.2f}"


def _bucket(customer_id: str) -> int:
    h = hashlib.sha256(str(customer_id).encode("utf-8")).digest()
    return int.from_bytes(h[:4], "big") % 100


def pick_psp(customer_id: str) -> str:
    canary = int(os.getenv("CANARY_PCT", "0"))
    return "paytm" if _bucket(customer_id) < canary else "razorpay"


def _paytm_base() -> str:
    return (
        "https://secure.paytmpayments.com"
        if os.getenv("PAYTM_ENVIRONMENT", "staging") == "production"
        else "https://securestage.paytmpayments.com"
    )


def _sanitize_cust_id(s: str) -> str:
    import re

    return re.sub(r"[^A-Za-z0-9_@-]", "_", str(s))


def create_razorpay_order(order_id: str, amount, customer_id: str) -> dict:
    client = razorpay.Client(auth=(os.getenv("RAZORPAY_KEY_ID"), os.getenv("RAZORPAY_KEY_SECRET")))
    paise = int(Decimal(str(amount)) * 100)
    res = client.order.create(
        {
            "amount": paise,
            "currency": "INR",
            "receipt": order_id,
            "notes": {"customer_id": customer_id},
        }
    )
    return {
        "psp": "razorpay",
        "psp_order_id": res["id"],
        "client_payload": {
            "key": os.getenv("RAZORPAY_KEY_ID"),
            "order_id": res["id"],
            "amount": res["amount"],
        },
    }


def create_paytm_order(order_id: str, amount, customer_id: str) -> dict:
    body = {
        "requestType": "Payment",
        "mid": os.getenv("PAYTM_MID"),
        "websiteName": os.getenv("PAYTM_WEBSITE_NAME", "WEBSTAGING"),
        "orderId": order_id,
        "callbackUrl": os.getenv("PAYTM_CALLBACK_URL"),
        "txnAmount": {"value": _money(amount), "currency": "INR"},
        "userInfo": {"custId": _sanitize_cust_id(customer_id)},
    }
    signature = PaytmChecksum.generateSignature(json.dumps(body), os.getenv("PAYTM_MERCHANT_KEY"))
    url = (
        f"{_paytm_base()}/theia/api/v1/initiateTransaction"
        f"?mid={os.getenv('PAYTM_MID')}&orderId={order_id}"
    )
    r = requests.post(
        url,
        json={"head": {"signature": signature}, "body": body},
        timeout=15,
    )
    js = r.json()
    token = (js.get("body") or {}).get("txnToken")
    if not token:
        msg = ((js.get("body") or {}).get("resultInfo") or {}).get("resultMsg", "no token")
        raise RuntimeError(f"Paytm initiate failed: {msg}")

    return {
        "psp": "paytm",
        "psp_order_id": order_id,
        "txn_token": token,
        "client_payload": {
            "mid": os.getenv("PAYTM_MID"),
            "orderId": order_id,
            "txnToken": token,
            "amount": _money(amount),
        },
    }


def create_order(order_id: str, amount, customer_id: str) -> dict:
    psp = pick_psp(customer_id)
    if psp == "paytm":
        out = create_paytm_order(order_id, amount, customer_id)
    else:
        out = create_razorpay_order(order_id, amount, customer_id)
    out["psp_routed"] = psp
    return out
